packages/utils/telemetry-utils/src/config.ts

```diff
@@ -348,3 +348,59 @@ export function createChildMonitoringContext(
 ): MonitoringContext {
 	return loggerToMonitoringContext(createChildLogger(props));
 }
+
+/**
+ * @internal
+ * */
+export type OptionConfigReaders<T extends object> = {
+	[K in keyof T]?: K extends string
+		? (config: IConfigProvider, name: `Fluid.${string}.${K}`) => T[K] | undefined
+		: undefined;
+};
+
+/**
+ * Creates a proxy object that allows for reading configuration values from a IConfigProviderBase,
+ * and default to the provided options if the configuration value is not present.
+ *
+ * @param config - the configuration provider to read values from.
+ * @param namespace - the namespace to use when reading configuration values.
+ * @param configReaders - a mapping of option keys to configuration value readers.
+ * @param defaultOptions - the default options to use if the configuration value is not present.
+ *
+ * @internal
+ * */
+export function createConfigBasedOptionsProxy<T extends object>(
+	config: IConfigProviderBase,
+	namespace: `Fluid.${string}`,
+	configReaders: OptionConfigReaders<T>,
```

Comment:
This seems to require listing out all of the expected options and calling `getBoolean`, etc on `realConfig`. Would it be possible to just automatically call `c.getRawConfig(n)` for any key or something so that we do not need to duplicate the declaration of options?

Reply:
This might lose some type safety / flexibility though. I'd just like to reduce the repetition of code like the following:
```typescript
				stringFeature: (c, n) => c.getString(n),
				arrayFeature: (c, n) => c.getNumberArray(n),
				booleanFeature: (c, n) => c.getBoolean(n),
				numberFeature: (c, n) => c.getNumber(n),
```

Maybe something like `stringFeature: "string", arrayFeature: "number[]"` etc. More sophisticated options such as `objectFeature` could still use a custom reader function, but trivial readers should be implementable using just a key from `ConfigTypeStringToType`

Reply:
What's funny is that was exactly what I had to start, but i found the API a bit awkward to use, and it was hard to get typing correct. There are also some cases where the config and option don't align 1:1, like we add a 2 to the config name, so that we have more control if we find a bug during a rollout, and what to try again, but not have old buggy clients turn the feature on. You can see this with summarizeProtocolTree in the container class.

Reply:
Yeah the reader functions are a necessary degree of freedom, but I think we could derive the type information from the reader object instead of manually declaring something like `IFeatureOptions`. This doesn't have to block the PR, but I think something like this could work:

```typescript
const optionReaders = {
  stringFeature: (c, n) => c.getString(n),
  arrayFeature: (c, n) => c.getNumberArray(n),
  booleanFeature: (c, n) => c.getBoolean(n),
  numberFeature: (c, n) => c.getNumber(n),
  objectFeature: (c, n) => {
    const str = c.getString(n);
    if (str !== undefined) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return JSON.parse(str);
    }
  },
  defaultedFeature: {
    default: 42
  }
}

type IFeatureOptions<TReaders> = {
  [K in keyof TReaders]: TReaders<K> extends (...args: any) => any
    ? ReturnType<TReaders<K>> // use the return type of the reader function
    : TReaders<K> extends { default: infer TDefault }
    ? TDefault
    : never;
}
```